django/core/servers/basehttp.py

```diff
@@ -97,6 +97,21 @@ def handle_error(self):
         if not is_broken_pipe_error():
             super().handle_error()
 
+    def finish_response(self):
+        try:
+            if self.environ['REQUEST_METHOD'] == 'HEAD':
+                # Consume the iterator for possible side effects
+                for content in self.result:
+                    pass
```

Comment:
This doesn't feel right.

Reply:
Do you have a better suggestions on how to consume an iterator without materializing it? That seems the most elegant way of doing it.

Reply:
@charettes I agree with you. Maybe this is better:
```
for _ in self.result:
    pass
```

Reply:
Perhaps the following?
```python
collections.deque(self.result, maxlen=0)  # consume iterator quickly.
```
Some performance numbers:
```
Python 3.7.0 (default, Sep 15 2018, 19:13:07) 
Type 'copyright', 'credits' or 'license' for more information
IPython 6.5.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from collections import deque
   ...: from itertools import repeat
   ...: n = 10000
   ...: %timeit -n1000 [__ for __ in repeat(0, n)]
   ...: %timeit -n1000 for __ in repeat(0, n): pass
   ...: %timeit -n1000 list(repeat(0, n))
   ...: %timeit -n1000 deque(repeat(0, n), maxlen=0)
195 µs ± 5.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
74.3 µs ± 204 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
53.2 µs ± 4.66 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
43 µs ± 428 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
So even calling `list()` seems better than an empty loop, and `collections.deque()` is fastest because it is written in C.

Reply:
You have to set up the iterator in each loop executed by `%timeit` otherwise it only consumes on the first loop after which it is empty...

I guess it depends on the size of the response being consumed. If you used `HEAD` on a response which was reasonably large, then the plain for loop is going to be slower.

Reply:
> You have to set up the iterator in each loop executed by `%timeit` otherwise it only consumes on the first loop after which it is empty...

I am not sure what you mean. I have to check the documentation for `%timeit`

> I guess it depends on the size of the response being consumed. If you used `HEAD` on a response which was reasonably large, then the plain for loop is going to be slower.

What about memory allocation?

```
In [21]: def large_body(n=100):
    ...:     return repeat(b'spam' * 1024 * 1024 * 5, n)
    ...:
In [24]: %timeit -n100 [__ for __ in large_body()]
5.8 ms ± 62.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
In [25]: %timeit -n100 for _ in large_body(): pass
5.94 ms ± 56.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
In [26]: %timeit -n100 list(large_body())
5.87 ms ± 63.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
In [27]: %timeit -n100 deque(large_body(), maxlen=0)
5.73 ms ± 91.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
In this case the elements all share the same storage (the string is the same for each iteration). If you construct a fresh string each time:
```
In [29]: def large_body(n=100):
    ...:     for _ in range(n):
    ...:         yield b'spam' * 1024 * 1024 * 5
    ...:
In [42]: %timeit [__ for __ in large_body(50)]
710 ms ± 9.39 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [43]: %timeit for __ in large_body(50): pass
277 ms ± 3.18 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [44]: %timeit list(large_body(50))
714 ms ± 3.32 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [45]: %timeit deque(large_body(50), maxlen=0)
279 ms ± 2.68 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
For _smaller_ but still large responses
```
In [46]: %timeit [__ for __ in large_body(10)]
54.1 ms ± 768 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [47]: %timeit for __ in large_body(10): pass
57 ms ± 865 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [48]: %timeit list(large_body(10))
52.6 ms ± 591 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [49]: %timeit deque(large_body(10), maxlen=0)
57.5 ms ± 479 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```